core/src/main/java/apoc/trigger/Trigger.java

```diff
@@ -18,41 +20,40 @@
  */
 
 public class Trigger {
-    public static class TriggerInfo {
-        public String name;
-        public String query;
-        public Map<String,Object> selector;
-        public Map<String, Object> params;
-        public boolean installed;
-        public boolean paused;
-
-        public TriggerInfo(String name, String query, Map<String, Object> selector, boolean installed, boolean paused) {
-            this.name = name;
-            this.query = query;
-            this.selector = selector;
-            this.installed = installed;
-            this.paused = paused;
-        }
+    // public for testing purpose
+    public static final String DB_NON_WRITER_ERROR = """
+            This instance is not allowed to write to the database because it is a follower.
+            Please use these procedures against the database leader.
+            """;
+    
+    @Context
+    public GraphDatabaseAPI db;
 
-        public TriggerInfo( String name, String query, Map<String,Object> selector, Map<String,Object> params, boolean installed, boolean paused )
-        {
-            this.name = name;
-            this.query = query;
-            this.selector = selector;
-            this.params = params;
-            this.installed = installed;
-            this.paused = paused;
-        }
-    }
+    @Context public TriggerHandler triggerHandler;
 
-    @Context public GraphDatabaseService db;
+    @Context public Log log;
 
-    @Context public TriggerHandler triggerHandler;
+    private void preprocessDeprecatedProcedures() {
```

Comment:
I think this function tries to do too much. I tries to log a warning saying the procedure is deprecated, and then it also does validates that we're allowed to write to the session database. 

I also think that the current logic is incomplete, because not only do we need to check that we're allowed to write to the session database, we also need to check that we're allowed to write to the system database.

I would recommend splitting organising the code around three functions:
- log warn message
- validate we're allowed to write to the session database
- validate we're allowed to write to the system database

Reply:
I also wonder whether the warning message is needed. I've noticed that these procedures are annotated with 

```java
@Deprecated(name = foo, deprecatedBy = bar)
```

I wonder whether this is in itself logging a deprecation warning in the database response? Iff this is true, then this would be even better than logging the deprecation to the stdout file and at that point I would probably recommend not logging the warning to the stdout file since I don't think the dbms does that.

Reply:
Nothing seems to be logged with `@Deprecated`, the only difference I notice is this (on mouse over)

<img width="775" alt="Schermata 2023-01-03 alle 12 25 20" src="https://user-images.githubusercontent.com/25103389/210378353-c35bb987-ac60-45d3-9c7c-7c17364533b0.png">

But for example in cypher-shell, there's no way to check at a glance whether it's deprecated.

I added the system db check as well.

Reply:
```json
{   "query": {     "text": "CALL apoc.warmup.run(false, false, false)",     "parameters": {}   },   "queryType": "r",   "counters": {     "_stats": {       "nodesCreated": 0,       "nodesDeleted": 0,       "relationshipsCreated": 0,       "relationshipsDeleted": 0,       "propertiesSet": 0,       "labelsAdded": 0,       "labelsRemoved": 0,       "indexesAdded": 0,       "indexesRemoved": 0,       "constraintsAdded": 0,       "constraintsRemoved": 0     },     "_systemUpdates": 0   },   "updateStatistics": {     "_stats": {       "nodesCreated": 0,       "nodesDeleted": 0,       "relationshipsCreated": 0,       "relationshipsDeleted": 0,       "propertiesSet": 0,       "labelsAdded": 0,       "labelsRemoved": 0,       "indexesAdded": 0,       "indexesRemoved": 0,       "constraintsAdded": 0,       "constraintsRemoved": 0     },     "_systemUpdates": 0   },   "plan": false,   "profile": false,   "notifications": [     {       "code": "Neo.ClientNotification.Statement.FeatureDeprecationWarning",       "title": "This feature is deprecated and will be removed in future versions.",       "description": "The query used a deprecated procedure. ('apoc.warmup.run' has been replaced by 'Firstly, the procedure duplicates functionality of page cache warm up which is a part of the DBMS. Secondly, the API of this procedure is very specific to Record storage engine.')",       "severity": "WARNING",       "position": {         "offset": 0,         "line": 1,         "column": 1       }     }   ],   "server": {     "address": "localhost:7687",     "agent": "Neo4j/5.1.0",     "protocolVersion": 5   },   "resultConsumedAfter": {     "low": 13,     "high": 0   },   "resultAvailableAfter": {     "low": 3,     "high": 0   },   "database": {     "name": "neo4j"   } }
```

The Neo4j DBMS has the ability to send back notifications in the query summary when returning query results to a user query. Whenever @Deprecated annotation is use, the Neo4j DBMS will send back a warning notification in the response for the overall query.

So I don't think we need to also be logging a warning message to the neo4j log file.

Reply:
_PS. This is what the notification looks like in the browser (it also shows the squiggly lines which you noticed in your screenshot)._

<img width="753" alt="Screenshot 2023-01-04 at 11 48 11" src="https://user-images.githubusercontent.com/9439499/210549564-62bd4cb5-7670-412b-af28-44d6175494c7.png">

Reply:
Btw, I made a mistake when I said we need to check that we need to do both:
- validate we're allowed to write to the session database
- validate we're allowed to write to the system database

I don't think we need to validate we're allowed to write to the session database, because both the new or the old procedures don't do that. Do they?

We only need to validate we're allowed to write to the system database, which is what the 4.4 behaviour does :+1:

Reply:
Yes, that makes sense. Deleted both log and session check